### ISSUE TITLE
Some questions about your paper

### ISSUE THREAD

**Issue by gaussleescorpio** [ThirstyScholar/trading-forex-with-reinforcement-learning#1]
Hi, Chien Yi,

I have read your paper with the title “financial trading as a game” recently. I would like to try your q learning model on currency data. After reading through your paper, I have some questions confusing me.

1. The reward function is calculated based on portfolio log return. However, in the action augmentation part, you also mentioned the reward is calculated for each action on every pair of currencies. Could you please tell me how the reward is exactly calculated? I am confused about this part. 

2. I think you are using one network to learn the trading models through the whole portfolio, right? So the q function is updated by samples from all the currency pairs at the same time, right? This is what I understand. Just wanna confirm with you. 

3. Do you have a source code regarding your project? If convenient, can I have the honor to read through?

**Comment by wert93** [comment]
Hello, I would also like to ask Chien Yi some questions about the paper. I will also try to answer gaussleescorpio's questions to the extent I have understood the paper so far. Perhaps if Chien Yi does not have time to reply we can figure out the implementation ourselves if you understand some of the stuff I haven't! 

About your issues: 
1. I believe that for each "augmented reward" he calculates how the portfolio would have changed according to those and based on that he simply gets the log portfolio return for each action at each step.

2. I intended it differently, as one separate network for each pair, because I thought it could be problematic, even with normalization, to associate such different currency pairs (like JPY and USD have very different ranges). But it might be like you say.

3. I would be very happy to take a look at the code too, and get directly the answers to all these questions! 

About my issues:

1. Reward calculation. I am not sure to understand how the timing of all the variables (action, old action, open and close prices) refer to each other. The way I got it from how it is written is that at each timestep (t) I receive from the environment the candle with OHLCV values and based on that I choose action A(t), which is then evaluated (getting the reward) on the same candle at timestep (t) on which I based my decision in the first place, and not, as I think it would be natural to do, at timestep (t+1) on the candle that still has to be received from the environment. But then you write clearly that the action is executed at the next open price, as it is realistic, so I do not really understand if I am just being very confused by the notation or if the action is really evaluated on the same step in which it is chosen. To be clear: A(t) refers to the action decided at timestep (t-1) and executed at open price of (t), or refers to the action decided at (t) and executed at (t+1)?

That is the main issue for me because, unfortunately, when I implement the reward function in the first way I get very good results (like yours and better), but when I implement it in the second, more natural, way I am almost always unable to find convergence to a local maximum and always get stuck with bad policies. Especially if I keep track of the actual ask and bid prices, and use those to calculate the portfolio so to have a more realistic spread cost. So in the first case, I believe that the setting is not realistically extensible to a live market; in the second case, my implementation is faulty and I need to revise it completely. 

2. Weight initialization: you say that you" initialize all hidden-to hidden weight matrices to be identity", - I am confused by the "all": is not there only one of those? Namely, the matrix between the second ELU layer and the LSTM. The first ELU layer only has a input-to hidden matrix and a bias matrix, right? 

3. Do you train the network on the whole sequence of 96 steps - so keeping the sequence of 96 Q-values, or only on the last value? I imagine you used pytorch, so I dont know about it, but in Keras you can choose if returning the whole sequence and I could not find an answer to that. 

4. Are the results showing the online algorithm also in the first steps of training? In this case, can you confirm that your algorithm gets to convergence in about 30-40k steps?

5. If, as it is understandable, you do not want to share the code, is it possible to see some graphs of the converged Q-values and the loss? So that it is easier to understand if I am getting something very wrong. 

Thanks a lot if you will take the time to answer all these questions!
